cmd.sh

```diff
@@ -57,7 +57,7 @@ cd /opt/nodemcu-firmware
 # EXTRA_CCFLAGS="-DBUILD_DATE=... AND -DNODE_VERSION=..." to make turned into an escaping/expanding nightmare for which
 # I never found a good solution
 if [ "$CAN_MODIFY_VERSION" = true ]; then
-  sed -i 's/\(NodeMCU [^"]*\)/\1 built with Docker provided by frightanic.com\\n\\tbranch: '"$BRANCH"'\\n\\tcommit: '"$COMMIT_ID"'\\n\\tSSL: '"$SSL"'\\n/g' app/include/user_version.h
+  sed -i '/#define NODE_VERSION[[:space:]]/ s/$/ " built with Docker provided by frightanic.com\\n\\tbranch: '"$BRANCH"'\\n\\tcommit: '"$COMMIT_ID"'\\n\\tSSL: '"$SSL"'\\n"/g' app/include/user_version.h
```

Comment:
The longer I think about this the more confused I get...don't get it. I see that it's correct, but why? Maybe the `sed`-ninja can explain that. 

- `/#define NODE_VERSION[[:space:]]/` is just the pattern to find the line to operate upon (I mean the one between the '/' is the pattern)
- ' ' is purely cosmetic to enhance legibility
- `s/` is the `sed` substitute
- and then? does the `$/` mean "substitute the EOL with what follows"?

In earlier NodeMCU versions [we had something like](https://github.com/nodemcu/nodemcu-firmware/blob/84c005e85c78b4382e68633ab61e2f4928c838f5/app/include/user_version.h#L9) `#define NODE_VERSION	"NodeMCU 1.5.4.1"` and the whitespace before the " (which GitHub mangles) is a \t.

The [2.2 upgrade](https://github.com/nodemcu/nodemcu-firmware/blob/master/app/include/user_version.h#L14) brings `#define NODE_VERSION		"NodeMCU " ESP_SDK_VERSION_STRING "." NODE_VERSION_XSTR(NODE_VERSION_INTERNAL)` with two \t before the first ".

On the current `dev` version this would produce

```
NodeMCU 2.2.0.0 built with Docker provided by frightanic.com
	branch: dev
	commit: b81963a86deec5f4dc25e7d9f3883914095cf9dd
	SSL: false
 build created on 2018-04-04 21:05
 powered by Lua 5.1.4 on SDK 2.2.1(cfd48f3)
```
which isn't so nice because of the "NodeMCU 2.2.0.0" vs. "SDK 2.2.1" but that's a firmware issue. Without this fix it would be

```
NodeMCU  built with Docker provided by frightanic.com
	branch: dev
	commit: b81963a86deec5f4dc25e7d9f3883914095cf9dd
	SSL: false
2.2.0.0 build created on 2018-04-04 23:17
 powered by Lua 5.1.4 on SDK 2.2.1(cfd48f3)
```

Reply:
Yes, you've decoded the `sed` correctly. :)

Reply:
Hurray! This in turn means that the new expression is not backwards compatible.

Reply:
Hmmm, I guess not being a C developer I don't fully understand constraints of the `.h` syntax.

```
/tmp > echo '#define NODE_VERSION "NodeMCU 1.5.4.1"' |sed '/#define NODE_VERSION[[:space:]]/ s/$/ " built with Docker provided by frightanic.com"/g'
#define NODE_VERSION "NodeMCU 1.5.4.1" " built with Docker provided by frightanic.com"
```

The result looks fishy because of the `" " `. Or is this automatically concatenated at runtime?

Reply:
Apparently then, two string literals separated by a space are still considered adjacent. There I was thinking the string concatenation _in PHP and Lua_ was obscure... Thanks for enlightening me 😜

Reply:
Mm, making this more (or less, depending on your view) exciting, this is strictly a compile-time thing.  You can't concatenate strings at runtime this way.  In any case, we good to merge this PR? :)

Reply:
Sorry, had to rush before I got around to actually push the button.

Reply:
No worries! :+1: